source/unified-test-format/unified-test-format.rst

```diff
@@ -1798,6 +1843,134 @@ Use a ``listIndexes`` command to check whether the index exists. Note that it is
+    when the failure was encountered.
+
+- ``storeSuccessesAsEntity``: if specfied, the runner MUST keep track of
+  the number of sub-operations that completed successfully, and store
+  that number in the specified entity. For example, if the loop contains
+  two sub-operations, and they complete successfully, each loop execution
+  would increment the number of successes by 2.
+
+- ``storeIterationsAsEntity``: if specfied, the runner MUST keep track of
+  the number of iterations of the loop performed, and store that number
+  in the specified entity.
+
+A *failure* is when an operation performed by the test runner has a
+different result from what the test expected. For example, a test
+containing an expectation that a particular document has a particular value
+in a particular field, and the actual value being different, would
+be a failure. An *error* is any other type of error. For example,
+attempting to invoke an undefined operation would be an error.
+
+This specification permits the test runner to report some failures as errors
+and some errors as failures. When the test runner stores errors and
+failures as entities it MAY classify conditions as errors and failures in
+the same way as it would when used in the driver's test suite.
+This includes reporting all errors as failures or all failures as errors.
+
+If the test runner does not distinguish errors and failures in its reporting,
+it MAY report both conditions under either category, but it MUST report
+any given condition in at most one category.
+
+The same entity name MAY be used for both ``storeErrorsAsEntity``
+and ``storeFailuresAsEntity`` options. The test runner MUST support
+writing both errors and failures to the same entity. If the same entity is
+specified, the errors and failures MUST be written in chronological order
+to the extent possible, interleaving the error and failure entries
+if necessary. Note that the textual description field for both errors
+and failures is ``error``.
+
+The following termination behavior MUST be implemented by the test
+runner:
+
+- The test runner MUST provide a way to request termination of loops. This
+  request will be made by the Atlas testing workload executor in response to
+  receiving the termination signal from Astrolabe.
+  
+- When the termination request is received, the test runner MUST
+  stop looping. If the test runner is looping when the termination request
+  is received, the current loop iteration MUST complete to its natural
+  conclusion (success or failure). If the test runner is not looping
+  when the termination request is received, it MUST NOT start any new
+  loop iterations in either the current test or subsequent tests for the
+  lifetime of the test runner; the test runner MAY skip any non-loop
+  operations and terminate as soon as is practical, or it MAY execute
+  any non-loop operations until either a loop is encountered or the
+  list of operations to execute is exhausted.
```

Comment:
I don't feel strongly about this, but to anticipate feedback others may have: is there any reason _not_ to be prescriptive about whether the test runner is allowed to continue executing operations? If there's not a technical challenge, wouldn't it be preferable for all drivers to guarantee the same behavior?

The termination itself does not directly influence success/failure; however, allowing flexibility here could lead to the same test producing different results in different drivers. Consider a test with the following operations:

 * `loop` with one operation, which inserts a document into a collection
 * `aggregate` that counts documents in the collection and emits a single result document based on the number of documents previously inserted (e.g. `{ over9000: <boolean> }`). This operation has an `expectResult` assertion (e.g. `{ over9000: true }`).

Possible behavior:

 * If drivers terminate by only prohibiting future loop iterations but not subsequent non-loop operations, `aggregate` may pass or fail depending on how long the loop ran before terminating.
 * If drivers terminate by aborting the loop _and_ skipping subsequent operations in the same test, the test is likely to always succeed
 * Suppose the `expectResult` in the post-loop operation was written to always fail. The test could still pass depending on whether drivers decide to skip non-loop operations after a loop termination signal.

Reply:
This is a good point and I think there may be a use case for post-loop operations for cleanup. I changed this section to mandate running all non-loop operations with a soft prohibition on the test runner containing post-loop operations whose success depends on the number of loop iterations performed, what do you think?

Reply:
New text looks good and addresses my concern above. Resolving.